lib/puppet/settings/config_file.rb

```diff
@@ -25,7 +25,6 @@ def parse_file(file, text, allowed_section_names = [])
       allowed_section_names << 'main' unless allowed_section_names.include?('main')
     end
 
-    # in Ruby 1.9.3 strings are not UTF-8 by default, so ensure text is treated properly
```

Comment:
Was there an associated code change that was meant to be here, too?

Reply:
I was thinking it's better to be explicit and continue to set the encoding, but ....

`Puppet::Settings#parse_file` calls `read_file` which always reads the file as [`UTF-8`](https://github.com/joshcooper/puppet/blob/270451074538750f00b71b83b42a68b6657e78d6/lib/puppet/settings.rb#L1220). The resulting text is passed to `Puppet::Settings::ConfigFile#parse_file`, and it just so happens that we set the `StringIO` encoding to the same encoding as the string it wraps... But if someone were to call `ConfigFile#parse_file` with a non-UTF-8 encoded string, then the `set_encoding` call would result in a UTF-8 string with embedded NULLs:

```
irb(main):086:0> hi = "hi".encode("UTF-16LE")
=> "hi"
irb(main):087:0> io = StringIO.new(hi)
=> #<StringIO:0x007fe19c537a98>
irb(main):088:0> io.set_encoding("UTF-8")
=> #<StringIO:0x007fe19c537a98>
irb(main):089:0> io.string.chars
=> ["h", "\u0000", "i", "\u0000"]
irb(main):090:0> io.string.bytes
=> [104, 0, 105, 0]
irb(main):091:0> io.string.encoding
=> #<Encoding:UTF-8>
```

I think it would be better to either call `set_encoding` with the string's encoding:

```
irb(main):092:0> hi = "hi".encode("UTF-16LE")
=> "hi"
irb(main):093:0> io = StringIO.new(hi)
=> #<StringIO:0x007fe19c4ec020>
irb(main):094:0> io.set_encoding(hi.encoding)
=> #<StringIO:0x007fe19c4ec020>
irb(main):095:0> io.string.chars
=> ["h", "i"]
irb(main):096:0> io.string.bytes
=> [104, 0, 105, 0]
irb(main):097:0> io.string.encoding
=> #<Encoding:UTF-16LE>
```

Or just drop the `set_encoding` call, as `StringIO` seems to derive the external encoding based on the string's encoding that it wraps:

```
irb(main):098:0> hi = "hi".encode("UTF-16LE")
=> "hi"
irb(main):099:0> io = StringIO.new(hi)
=> #<StringIO:0x007fe19d0eaed0>
irb(main):100:0> io.string.chars
=> ["h", "i"]
irb(main):101:0> io.string.bytes
=> [104, 0, 105, 0]
irb(main):102:0> io.string.encoding
=> #<Encoding:UTF-16LE>
```

But it results in a non-UTF-8 string. So the best thing is to transcode the string to "UTF-8" before wrapping it in a `StringIO`:

```
irb(main):107:0> hi = "hi".encode("UTF-16LE")
=> "hi"
irb(main):108:0> io = StringIO.new(hi.encode("UTF-8"))
=> #<StringIO:0x007fe19e17b150>
irb(main):109:0> io.string.chars
=> ["h", "i"]
irb(main):110:0> io.string.bytes
=> [104, 105]
irb(main):111:0> io.string.encoding
=> #<Encoding:UTF-8>
```

Reply:
Yeah, that makes sense. I'm 👍 for updating it to be the `StringIO.new(hi.encode("UTF-8"))` form.

Reply:
Yeah, so I'm trying to understand my original commit at https://github.com/puppetlabs/puppet/commit/b27d2117aeca713cf9e75658349b8f435e69bf32, which sadly did not add new tests.

```diff
-    ini = Puppet::Settings::IniFile.parse(StringIO.new(text))
+    # in Ruby 1.9.3 strings are not UTF-8 by default, so ensure text is treated properly
+    ini = Puppet::Settings::IniFile.parse(StringIO.new(text).set_encoding(Encoding::UTF_8))
```

I will speculate that was done because the file is parsed as UTF-8, but the string `text` could be in another encoding - signature of method:

```ruby
def parse_file(file, text, allowed_section_names = [])
```

So I think intent was to make sure to search the UTF-8 string with a UTF-8 string, but as you've pointed out, this makes no sense based on `set_encoding` not performing transcoding.

What you've done looks correct to me.